pkg/sql/alter_column_type_test.go

```diff
@@ -0,0 +1,521 @@
+	defer s.Stopper().Stop(ctx)
+
+	if _, err := sqlDB.Exec(
+		`SET enable_alter_column_type_general = true;`); err != nil {
+		t.Error(err)
+	}
+
+	if _, err := sqlDB.Exec(`
+CREATE TABLE test (x INT);
+ALTER TABLE test ALTER COLUMN x TYPE STRING;
+INSERT INTO test VALUES ('1'), ('2'), ('3');
+`); err != nil {
+		t.Error(err)
+	}
+
+	var x string
+	dest := []interface{}{&x}
+
+	expected := [][]interface{}{{"1"}, {"2"}, {"3"}}
+	err := sql.QueryExpect(`SELECT * FROM test ORDER BY x`, sqlDB, expected, dest)
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+func TestQueryIntToString(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer setTestJobsAdoptInterval()()
+
+	params, _ := tests.CreateTestServerParams()
+
+	s, sqlDB, _ := serverutils.StartServer(t, params)
+	ctx := context.TODO()
+	defer s.Stopper().Stop(ctx)
+
+	if _, err := sqlDB.Exec(
+		`SET enable_alter_column_type_general = true;`); err != nil {
+		t.Error(err)
+	}
+
+	if _, err := sqlDB.Exec(`
+CREATE DATABASE d;
+CREATE TABLE d.t3 (x INT, y INT, z INT);
+INSERT INTO d.t3 VALUES (1, 1, 1), (2, 2, 2);
+ALTER TABLE d.t3 ALTER COLUMN y TYPE STRING;
+`); err != nil {
+		t.Error(err)
+	}
+
+	testutils.SucceedsSoon(t, func() error {
+		_, err := sqlDB.Exec(`INSERT INTO d.t3 VALUES (3, 'HELLO', 3);`)
+		return err
+	})
+
+	var y string
+	var x, z int
+	dest := []interface{}{&x, &y, &z}
+
+	expected := [][]interface{}{{1, "1", 1}, {2, "2", 2}, {3, "HELLO", 3}}
+	err := sql.QueryExpect(`SELECT * FROM d.t3 ORDER BY x`, sqlDB, expected, dest)
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+func TestAlterColumnTypeBeforeAlterPrimaryKey(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	ctx := context.Background()
+
+	swapNotification := make(chan struct{})
+	waitBeforeContinuing := make(chan struct{})
+
+	params, _ := tests.CreateTestServerParams()
+	params.Knobs = base.TestingKnobs{
+		SQLSchemaChanger: &sql.SchemaChangerTestingKnobs{
+			RunBeforeComputedColumnSwap: func() {
+				swapNotification <- struct{}{}
+				<-waitBeforeContinuing
+			},
+		},
+		DistSQL: &execinfra.TestingKnobs{},
+	}
+
+	s, sqlDB, _ := serverutils.StartServer(t, params)
+	defer s.Stopper().Stop(ctx)
+
+	if _, err := sqlDB.Exec(
+		`SET enable_alter_column_type_general = true;`); err != nil {
+		t.Error(err)
+	}
+
+	if _, err := sqlDB.Exec(`
+CREATE DATABASE t;
+CREATE TABLE t.test (x INT NOT NULL);
+`); err != nil {
+		t.Error(err)
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		if _, err := sqlDB.Exec(`
+ALTER TABLE t.test ALTER COLUMN x TYPE STRING;
+`); err != nil {
+			t.Error(err)
+		}
+		wg.Done()
+	}()
+
+	<-swapNotification
+
+	expected := "pq: unimplemented: table test is currently undergoing a schema change"
+	_, err := sqlDB.Exec(`ALTER TABLE t.test ALTER PRIMARY KEY USING COLUMNS (x);`)
+	waitBeforeContinuing <- struct{}{}
+	wg.Wait()
+	if err.Error() != expected {
+		t.Fatalf("expected error: %s, got: %s", expected, err.Error())
+	}
+}
+
+func TestAlterPrimaryKeyBeforeAlterColumnType(t *testing.T) {
```

Comment:
I'd rename this to something like TestSchemaChangeBeforeAlterColumnType -- theres nothing unique about the primary key change here right?